src/couch_stats/src/couch_stats_histogram.erl

```diff
@@ -0,0 +1,449 @@
+%  <-1-> <-----------14------------> <--49--->
+%        ^^^^^^^ bin index ^^^^^^^^^
+%
+% With Erlang's wonderful binary matching capabilities this becomes a
+% one-liner:
+%
+%    <<0:1, BinIndex:14, _/bitstring>> = <<Val/float>>
+%
+% The internal implementation is a tuple of counter:new(?BIN_COUNT)
+% elements. The tuple size is determined by the time window parameter when the
+% histogram is created. After a histogram object is created, its Erlang term
+% structure is static so it's suitable to be stored in a persistent term [3].
+% The structure looks something like:
+%
+%  {
+%     1          = [1, 2, ..., ?BIN_COUNT]
+%     2          = [1, 2, ..., ?BIN_COUNT]
+%     ...
+%     TimeWindow = [1, 2, ..., ?BIN_COUNT]
+%  }
+%
+% The representation can also be regarded as a table with rows as abstract time
+% units: 1 = 1st second, 2 = 2nd second, etc, and the columns as histogram
+% bins. The update/3 function takes a value and a current time as parameters.
+% The time is used to select which row to update, and the value picks which
+% bin to increment.
+%
+% In practice, the time window would be used as a circular buffer. The time
+% parameter to the update/3 function might be the system monotonic clock time
+% and then the histogram time index is compuated as `Time rem TimeWindow`. So,
+% as the monotonic time is advancing forward, the histogram time index will
+% loop around. This comes with a minor annoynance of having to allocate a
+% larger time window to accomodate some process which cleans stale (expired)
+% histogram entries, possibly with some extra buffers to ensure the currently
+% updated interval and the interval ready to be cleaned would not overlap.
+%
+% Reading a histogram can be done via the read/3 function. The function takes a
+% start time and an interval. All the histogram entries in the interval will be
+% summed together, bin by bin, and returned as a new counters object. This
+% functionality can be used to gather and merge multiple histogram together.
+%
+% To get a stats summary across a time window use the stats/3 function. Just
+% like the read/3 function, it takes a start time and a time interval over
+% which to summarize the data.
+%
+% In addition to the new/1, update/3, read/3, stats/3 there are simple
+% functions which default to using WindowSize = 1. The intent is they would be
+% used when a simple histogram is needed without the time window functionality.
+%
+% [1] https://www.erlang.org/doc/man/counters.html
+% [2] https://en.wikipedia.org/wiki/IEEE_754
+% [3] https://www.erlang.org/doc/man/persistent_term.html
+
+-module(couch_stats_histogram).
+
+-export([
+    new/0,
+    new/1,
+
+    update/2,
+    update/3,
+
+    stats/1,
+    stats/3,
+
+    read/1,
+    read/3,
+
+    clear/1,
+    clear/3,
+
+    bin_min/1,
+    bin_max/1,
+    bin_middle/1,
+
+    calc_bin_offset/0,
+    calc_bin_count/0,
+    get_bin_boundaries/0
+]).
+
+-on_load(check_constants/0).
```

Comment:
Ah good point about having both `-on_load(...)` handler and `calc_*` functions. I added `-on_load(...)` later but didn't consider that it would prevent using the `calc_*` functions.

Wonder if we should just rely on eunit tests and skip `-on_load(...)` or maybe just add a comment by `-on_load(...)` specifying to comment it out when updating constants?

Reply:
> I could see an argument for the minimum value be one microsecond, so I decided to see how things change if we drop MIN_VAL to 0.001. BIN_COUNT increases to 256, which is maybe too much memory overhead for such a small increase in low value resolution.

I started with 1 microsecond but noticed that above a certain counters size (232) there was slight jump in update speed, probably due to some cache line size limit. It was a small difference but there was also the additional memory overhead so that's how I ended up with 10 microseconds. For most request times that seems like a decent minimum.

Reply:
Discussing memory overhead looked a bit at how much memory these histogram are using.

On a 12 scheduler Erlang 24 VM:

```erlang
> f(H), H = couch_stats_histogram:new(30).
> f(Mem), #{memory := Mem} = counters:info(element(1, H)), Mem * 30.
724800
```

It's about 700KB. Since we're using `write_concurrency` mode, the counters are striped (hashed) across all schedulers, there is a per-scheduler overhead here too.

Using an atomic (default) counter memory usage is lower.
```erlang
f(Mem), #{memory := Mem} = counters:info(counters:new(230, [])), Mem * 30.
56400.
```

However in the presence of concurrency we'd want the `write_concurrency` variant. If we had a per-process or per-shard type histograms the atomic variant could be used.

A Folsom histogram for comparison is about 1.4MB

```erlang
> ets:info(#Ref<0.4093066570.1809186818.69626>, name).
 folsom_slide_uniform
> ets:info(#Ref<0.4093066570.1809186818.69626>, memory) * erlang:system_info(wordsize).
1471800
```